frame/scheduler/src/lib.rs

```diff
@@ -92,13 +99,36 @@ pub struct Scheduled<Call, BlockNumber> {
 	call: Call,
 	/// If the call is periodic, then this points to the information concerning that.
 	maybe_periodic: Option<schedule::Period<BlockNumber>>,
+	/// The origin to dispatch the call.
+	origin: PalletsOrigin,
+	_phantom: PhantomData<AccountId>,
+}
+
+impl<Call, BlockNumber, PalletsOrigin, AccountId> Decode for Scheduled<Call, BlockNumber, PalletsOrigin, AccountId>
+where
+	Call: Decode,
+	BlockNumber: Decode,
+	PalletsOrigin: Decode + From<system::RawOrigin<AccountId>>
+{
+	fn decode<I: codec::Input>(input: &mut I) -> sp_std::result::Result<Self, codec::Error> {
+		// Implement decode manually to keep backwards compatibility.
+		// Old version does not have origin field and default to ROOT.
+		Ok(Scheduled {
+			maybe_id: Decode::decode(input)?,
+			priority: Decode::decode(input)?,
+			call: Decode::decode(input)?,
+			maybe_periodic: Decode::decode(input)?,
+			origin: Decode::decode(input).unwrap_or_else(|_| system::RawOrigin::Root.into()),
+			_phantom: Default::default(),
+		})
+	}
```

Comment:
I think this is not valid because `Decode::decode(input).unwrap_or_else(|_| system::RawOrigin::Root.into()),` modifies input in case no origin.

Thus `Vec<Scheduled (without rigin)>` will remove some bytes of the second element when decoding the first one.

I think a runtime upgrade to upgrade the storage in scheduler and an explicit Message that other chain must upgrade as well if they store `Scheduled` in the PR title should be fine.

Reply:
Migration code added. The pallet should handle the upgrade automatically so I don't think pallet users needs to do any action and we don't need explicit message?

Reply:
I was thinking because `Scheduled` implements `Encode/Decode` and is public some user of scheduler could have some instance in their pallet but indeed this is very unlikely so I don't know, I guess its fine